registry/aws.go

```diff
@@ -0,0 +1,216 @@
+		if c := awsCreds.credsFor(domain); c == (creds{}) {
+			goto refresh
+		}
+
+		// otherwise, check if the tokens have expired
+		if expiry, ok := regionExpire[region]; !ok || expiry.Before(now) {
+			goto refresh
+		}
+
+		// the creds exist and are before the use-by; nothing to be done.
+		return nil
+
+	refresh:
+		// unconditionally append the sought-after account, and let
+		// the AWS API figure out if it's a duplicate.
+		accountIDs := append(allAccountIDsInRegion(awsCreds.Hosts(), region), accountID)
+		logger.Log("info", "attempting to refresh auth tokens", "region", region, "account-ids", strings.Join(accountIDs, ", "))
+		regionCreds, expiry, err := fetchAWSCreds(region, accountIDs)
+		if err != nil {
+			regionEmbargo[region] = now.Add(embargoDuration)
+			logger.Log("error", "fetching credentials for AWS region", "region", region, "err", err, "embargo", embargoDuration)
+			return err
+		}
+		regionExpire[region] = expiry
+		awsCreds.Merge(regionCreds)
+		return nil
+	}
+
+	return func() ImageCreds {
+		imageCreds := lookup()
+
+		for name, creds := range imageCreds {
+			domain := name.Domain
+			if strings.HasSuffix(domain, ecrHostSuffix) {
+				bits := strings.Split(domain, ".")
+				if len(bits) != 6 {
+					logger.Log("warning", "AWS registry domain not in expected format <account-id>.dkr.ecr.<region>.amazonaws.com", "domain", domain)
+					continue
+				}
+				accountID := bits[0]
+				region := bits[3]
+
+				if !shouldScan(region, accountID) {
+					delete(imageCreds, name)
+					continue
+				}
+				if err := ensureCreds(domain, region, accountID, time.Now()); err != nil {
+					logger.Log("warning", "unable to ensure credentials for ECR", "domain", domain, "err", err)
+				}
+				newCreds := NoCredentials()
+				newCreds.Merge(awsCreds)
+				newCreds.Merge(creds)
+				imageCreds[name] = newCreds
+			}
+		}
+		return imageCreds
+	}, nil
+}
+
+func allAccountIDsInRegion(hosts []string, region string) []string {
+	var ids []string
+	// this returns a list of unique accountIDs, assuming that the input is unique hostnames
+	for _, host := range hosts {
+		bits := strings.Split(host, ".")
+		if len(bits) != 6 {
+			continue
+		}
+		if bits[3] == region {
+			ids = append(ids, bits[0])
+		}
+	}
+	return ids
+}
+
+func fetchAWSCreds(region string, accountIDs []string) (Credentials, time.Time, error) {
+	sess := session.Must(session.NewSession(&aws.Config{Region: aws.String(region)}))
+	svc := ecr.New(sess)
+	ecrToken, err := svc.GetAuthorizationToken(&ecr.GetAuthorizationTokenInput{
+		RegistryIds: aws.StringSlice(accountIDs),
+	})
+	if err != nil {
+		return Credentials{}, time.Time{}, err
+	}
+	auths := make(map[string]creds)
+	expiry := time.Now().Add(defaultTokenValid)
+	for _, v := range ecrToken.AuthorizationData {
+		// Remove the https prefix
+		host := strings.TrimPrefix(*v.ProxyEndpoint, "https://")
+		creds, err := parseAuth(*v.AuthorizationToken)
+		if err != nil {
+			return Credentials{}, time.Time{}, err
```

Comment:
What will happen if we fail to refresh, when will it retry?

Reply:
I was looking in the wrong place, there are plenty of comments outside of this function.